src/main/java/de/zalando/aruha/nakadi/service/subscription/ExactWeightRebalancer.java

```diff
@@ -0,0 +1,113 @@
+package de.zalando.aruha.nakadi.service.subscription;
+
+import de.zalando.aruha.nakadi.service.subscription.model.Partition;
+import de.zalando.aruha.nakadi.service.subscription.model.Session;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.BiFunction;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+import java.util.stream.Stream;
+
+class ExactWeightRebalancer implements BiFunction<Session[], Partition[], Partition[]> {
+    @Override
+    public Partition[] apply(final Session[] sessions, final Partition[] currentPartitions) {
+        final Map<String, Session> activeClients = Stream.of(sessions).collect(Collectors.toMap(c -> c.id, c -> c));
+        // sorted client ids.
+        final List<String> activeClientIds = activeClients.keySet().stream().sorted().collect(Collectors.toList());
+        // the main part of rebalance - calculate count for each partition.
+        final int[] partitionsPerClient = splitByWeight(
+                currentPartitions.length,
+                activeClientIds.stream().mapToInt(c -> activeClients.get(c).weight).toArray());
+
+        // Stage 1. Select partitions that are not assigned to any EXISTING client.
+        final Set<Partition> unassignedPartitions = Stream.of(currentPartitions)
+                .filter(p -> Partition.State.UNASSIGNED.equals(p.getState()) || !activeClients.containsKey(p.getSessionOrNextSession()) || !activeClients.containsKey(p.getSession()))
```

Comment:
I had a hard time to understand all 3 conditions. Maybe could we have some intermediary steps better describing each condition?